File: go/apidef.go
/* Copyright (c) 2024-2025 Voxgig, MIT License */

package apidef

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	jostraca "github.com/jostraca/jostraca/go"
)

const VERSION = "0.11.0"

// ApiDef creates a new API definition generator with the given options.
func NewApiDef(opts ApiDefOptions) *apiDefInstance {
	if opts.Strategy == "" {
		opts.Strategy = "heuristic01"
	}
	return &apiDefInstance{
		opts: opts,
	}
}

type apiDefInstance struct {
	opts ApiDefOptions
}

// Generate processes an API definition spec and produces the output model.
func (a *apiDefInstance) Generate(spec map[string]any) (*ApiDefResult, error) {
	start := time.Now().UnixMilli()
	steps := make([]string, 0)

	// Build control from spec
	ctrl := map[string]any{
		"step": map[string]any{
			"parse":        true,
			"guide":        true,
			"transformers": true,
			"builders":     true,
			"generate":     true,
		},
	}
	if specCtrl, ok := spec["ctrl"].(map[string]any); ok {
		if stepCtrl, ok := specCtrl["step"].(map[string]any); ok {
			ctrlStep := ctrl["step"].(map[string]any)
			for _, k := range sortedKeys(stepCtrl) {
				ctrlStep[k] = stepCtrl[k]
			}
		}
	}

	ctrlStep := ctrl["step"].(map[string]any)

	// Initialize API model
	apimodel := map[string]any{
		"main": map[string]any{
			KIT: map[string]any{
				"info":   map[string]any{},
				"entity": map[string]any{},
				"flow":   map[string]any{},
			},
		},
	}

	// Build spec
	buildSpec := map[string]any{}
	if build, ok := spec["build"].(map[string]any); ok {
		if bs, ok := build["spec"].(map[string]any); ok {
			buildSpec = bs
		}
	}

	// Model
	model := map[string]any{}
	if m, ok := spec["model"].(map[string]any); ok {
		model = m
	}

	// Step: parse
	if parseBool, ok := ctrlStep["parse"].(bool); !ok || !parseBool {
		return &ApiDefResult{OK: true, Steps: steps, Start: start, End: time.Now().UnixMilli(), Ctrl: ctrl}, nil
	}

	SetCustomPlurals(modelCustomPlurals(model))
	defer ClearCustomPlurals()

	defName, _ := model["def"].(string)
	base, _ := buildSpec["base"].(string)
	defpath := filepath.Join(base, "..", "def", defName)

	// Create context
	warn := MakeWarner("warning", nil)
	ctx := &ApiDefContext{
		Spec:     spec,
		Opts:     a.opts,
		DefPath:  filepath.Dir(defpath),
		Model:    model,
		ApiModel: apimodel,
		Guide:    map[string]any{},
		Note:     map[string]any{},
		Warn:     warn,
		Metrics: &Metrics{
			Count: MetricsCount{
				OrigCmpRefs: map[string]int{},
			},
			Found: map[string]any{
				"cmp": map[string]any{},
				"tag": map[string]any{},
			},
		},
		Work: map[string]any{},
	}

	fail := func(err error) (*ApiDefResult, error) {
		warn.Warn(map[string]any{
			"err":  err.Error(),
			"note": "!! BUILD FAILED !! " + err.Error(),
		})
		WriteFileWarn(warn, "./apidef-warnings.txt", warningsFileText(warn.History()))
		return makeErrorResult(start, steps, ctrl, ctx, err), err
	}

	// Load and parse definition
	defsrc, err := LoadFile(defpath)
	if err != nil {
		return fail(err)
	}

	def, err := Parse("OpenAPI", defsrc, map[string]string{"file": defpath})
	if err != nil {
		return fail(err)
	}

	// Write debug file if debug mode
	if a.opts.Debug != nil {
		fullJSON, _ := json.MarshalIndent(def, "", "  ")
		os.WriteFile(defpath+".full.json", fullJSON, 0644)
	}

	ctx.Def = def
	buildctx, _ := spec["buildctx"].(map[string]any)
	config, _ := spec["config"].(map[string]any)
	kind, _ := config["kind"].(string)
	if kind == "" {
		kind = "openapi3"
	}
	ctx.Resolved = PublishResolved(buildctx, kind, def, func() map[string]any { return ctx.Guide })
	steps = append(steps, "parse")

	// Step: guide
	if guideBool, ok := ctrlStep["guide"].(bool); !ok || !guideBool {
		return &ApiDefResult{OK: false, Steps: steps, Start: start, End: time.Now().UnixMilli(), Ctrl: ctrl, Ctx: ctx}, nil
	}

	guideModel, err := BuildGuide(ctx)
	if err != nil {
		return fail(err)
	}
	if guideModel == nil {
		return fail(fmt.Errorf("unable to build guide"))
	}

	ctx.Guide, _ = guideModel["guide"].(map[string]any)
	steps = append(steps, "guide")

	// Step: transformers
	if transBool, ok := ctrlStep["transformers"].(bool); !ok || !transBool {
		return &ApiDefResult{
			OK: true, Steps: steps, Start: start, End: time.Now().UnixMilli(),
			Ctrl: ctrl, Guide: ctx.Guide, ApiModel: ctx.ApiModel, Ctx: ctx,
		}, nil
	}

	transforms := []Transform{
		TopTransform,
		EntityTransform,
		OperationTransform,
		ContractTransform,
		ArgsTransform,
		SelectTransform,
		FieldTransform,
		FlowTransform,
		FlowstepTransform,
		CleanTransform,
	}

	for _, t := range transforms {
		if _, err := t(ctx); err != nil {
			return fail(err)
		}
	}
	steps = append(steps, "transformers")

	// Step: builders
	if builderBool, ok := ctrlStep["builders"].(bool); !ok || !builderBool {
		return &ApiDefResult{
			OK: true, Steps: steps, Start: start, End: time.Now().UnixMilli(),
			Ctrl: ctrl, Guide: ctx.Guide, ApiModel: ctx.ApiModel, Ctx: ctx,
		}, nil
	}

	entityBuilder, err := MakeEntityBuilder(ctx)
	if err != nil {
		return fail(err)
	}
	flowBuilder, err := MakeFlowBuilder(ctx)
	if err != nil {
		return fail(err)
	}
	steps = append(steps, "builders")

	// Step: generate
	if genBool, ok := ctrlStep["generate"].(bool); !ok || !genBool {
		return &ApiDefResult{
			OK: true, Steps: steps, Start: start, End: time.Now().UnixMilli(),
			Ctrl: ctrl, Guide: ctx.Guide, ApiModel: ctx.ApiModel, Ctx: ctx,
		}, nil
	}

	builders := []Builder{entityBuilder, flowBuilder}

	jopts := []jostraca.Option{jostraca.WithLog(jostracaLog{ctx.Log})}
	if now, ok := spec["now"].(func() int64); ok {
		jopts = append(jopts, jostraca.WithNow(now))
	}
	write, merge := true, false

	jres, err := jostraca.New(jopts...).Generate(jostraca.Options{
		Folder: a.opts.Folder,
		Model:  map[string]any{},
		Existing: jostraca.Existing{
			Txt: jostraca.ExistingTxt{Write: &write, Merge: &merge},
		},
	}, func(j *jostraca.J) {
		j.Project(jostraca.ProjectProps{Folder: "."}, func(j *jostraca.J) {
			for _, builder := range builders {
				builder(j)
			}
		})
	})
	if err != nil {
		return fail(err)
	}

	steps = append(steps, "generate")

	kitEntity, _ := getKit(ctx)["entity"].(map[string]any)
	GcEntityFiles(ctx.Log, a.opts.Folder, a.opts.OutPrefix, sortedKeys(kitEntity))

	if warnings := warn.History(); len(warnings) > 0 {
		WriteFileWarn(warn, "./apidef-warnings.txt", warningsFileText(warnings))
	}

	return &ApiDefResult{
		OK:       true,
		Reload:   len(jres.Files.Written) > 0 || len(jres.Files.Merged) > 0,
		Start:    start,
		End:      time.Now().UnixMilli(),
		Steps:    steps,
		Ctrl:     ctrl,
		Guide:    ctx.Guide,
		ApiModel: ctx.ApiModel,
		Ctx:      ctx,
		Jres:     &jres,
	}, nil
}

func warningsFileText(history []map[string]any) string {
	texts := make([]string, 0, len(history))
	for _, w := range history {
		texts = append(texts, FormatJSONIC(w))
	}
	return strings.Join(texts, "\n\n")
}

// MakeBuild creates a build function from options.
func MakeBuild(opts ApiDefOptions) func(model, build map[string]any) (*ApiDefResult, error) {
	var apidef *apiDefInstance

	return func(model, build map[string]any) (*ApiDefResult, error) {
		if apidef == nil {
			apidef = NewApiDef(opts)
		}

		ctrl := map[string]any{}
		if ba, ok := build["spec"].(map[string]any); ok {
			if ba2, ok := ba["buildargs"].(map[string]any); ok {
				if ad, ok := ba2["apidef"].(map[string]any); ok {
					if c, ok := ad["ctrl"].(map[string]any); ok {
						ctrl = c
					}
				}
			}
		}

		return apidef.Generate(map[string]any{
			"model":    model,
			"build":    build,
			"buildctx": build,
			"ctrl":     ctrl,
		})
	}
}

func modelCustomPlurals(model map[string]any) any {
	main, _ := model["main"].(map[string]any)
	custom, _ := main["custom"].(map[string]any)
	if custom == nil {
		return nil
	}
	return custom["plurals"]
}

func makeErrorResult(start int64, steps []string, ctrl map[string]any, ctx *ApiDefContext, err error) *ApiDefResult {
	return &ApiDefResult{
		OK:       false,
		Err:      err,
		Start:    start,
		End:      time.Now().UnixMilli(),
		Steps:    steps,
		Ctrl:     ctrl,
		Guide:    ctx.Guide,
		ApiModel: ctx.ApiModel,
		Ctx:      ctx,
	}
}

// jostracaLog hands jostraca's replayed warnings to apidef's logger, as the
// TS port passes its own. Without one they are dropped: jostraca's default
// would print them to stdout.
type jostracaLog struct{ log Logger }

func (l jostracaLog) Trace(args ...any) { l.forward(Logger.Debug, args) }
func (l jostracaLog) Debug(args ...any) { l.forward(Logger.Debug, args) }
func (l jostracaLog) Info(args ...any)  { l.forward(Logger.Info, args) }
func (l jostracaLog) Warn(args ...any)  { l.forward(Logger.Warn, args) }
func (l jostracaLog) Error(args ...any) { l.forward(Logger.Error, args) }
func (l jostracaLog) Fatal(args ...any) { l.forward(Logger.Error, args) }

func (l jostracaLog) forward(level func(Logger, ...any), args []any) {
	if l.log != nil {
		level(l.log, args...)
	}
}

File: ts/src/apidef.ts
/* Copyright (c) 2024-2025 Voxgig, MIT License */


import * as Fs from 'node:fs'
import Path from 'node:path'


import {
  Jostraca, JostracaResult, Project, names
} from 'jostraca'

import { prettyPino } from '@voxgig/util'


import type {
  ApiDefOptions,
  ApiDefResult,
  Control,
  DefKind,
  Model,
  Build,
  ApiModel,
} from './types'


import {
  KIT,

  OpenModelShape,
  OpenBuildShape,
  OpenControlShape,
  ApiDefContext,
} from './types'


import {
  buildGuide,
} from './guide/guide'


import {
  parse,
} from './parse'


import {
  fixName,
} from './transform'


import {
  nom,
  loadFile,
  getdlog,
  makeWarner,
  formatJSONIC,
  depluralize,
  setCustomPlurals,
  clearCustomPlurals,
  sanitizeSlug,
  slugToPascalCase,
  writeFileSyncWarn,
  relativizePath,
  getModelPath,
  VALID_CANON,
  CANON_ONE,
} from './utility'


import { topTransform } from './transform/top'
import { entityTransform } from './transform/entity'
import { operationTransform } from './transform/operation'
import { graphqlTransform } from './transform/graphql'
import { argsTransform } from './transform/args'
import { selectTransform } from './transform/select'
import { fieldTransform } from './transform/field'
import { flowTransform } from './transform/flow'
import { flowstepTransform } from './transform/flowstep'
import { cleanTransform } from './transform/clean'

import { makeEntityBuilder } from './builder/entity'
import { gcEntityFiles } from './builder/entity/entity'
import { makeFlowBuilder } from './builder/flow'

// Log non-fatal wierdness.
const dlog = getdlog('apidef', __filename)


function ApiDef(opts: ApiDefOptions) {

  // TODO: shape opts!
  const fs = opts.fs || Fs
  const pino = prettyPino('apidef', opts)
  const log = pino.child({ cmp: 'apidef' })
  const warn = makeWarner({ point: 'warning', log })

  // Input format: explicit option wins, else sniff from the def file name.
  // A GraphQL build defaults to the graphql01 guide strategy.
  opts.kind = opts.kind || resolveKind(opts.def)
  opts.strategy = opts.strategy ||
    ('GraphQL' === opts.kind ? 'graphql01' : 'heuristic01')

  async function generate(spec: any): Promise<ApiDefResult> {
    const start = Date.now()
    const steps: string[] = []

    let ctx: ApiDefContext | undefined = undefined
    let ctrl: Control | undefined = undefined
    let jres: JostracaResult | undefined = undefined

    try {
      ctrl = OpenControlShape(spec.ctrl || {}) as Control

      const model: Model = OpenModelShape(spec.model || {})
      const build: Build = OpenBuildShape(spec.build || {})

      // Step: parse (API spec).
      if (!ctrl.step.parse) {
        return { ok: true, steps, start, end: Date.now(), ctrl }
      }

      names(model, model.name)

      // Install per-model plural overrides for depluralize/canonize.
      // Read from model.main.custom.plurals; downstream transforms
      // and the guide pick this up implicitly via the utility module.
      setCustomPlurals((model as any)?.main?.custom?.plurals)

      const apimodel: ApiModel = {
        main: {
          [KIT]: {
            info: {},
            entity: {},
            flow: {},
          },
        },
      }

      const buildspec = build.spec

      let defpath = model.def

      // TOOD: defpath should be independently defined
      defpath = Path.join(buildspec.base, '..', 'def', defpath)

      log.info({
        point: 'generate-start',
        note: relativizePath(defpath),
        defpath,
        start
      })

      // TODO: Validate spec
      ctx = {
        fs,
        fsInjected: null != opts.fs,
        log,
        spec,
        opts,
        util: { fixName },
        defpath: Path.dirname(defpath),
        model,
        apimodel,
        guide: {},
        def: undefined,
        note: {},
        warn,

        // TODO: remove (moved to guide)
        metrics: {
          count: {
            path: 0,
            method: 0,
            origcmprefs: {},
            cmp: 0,
            tag: 0,
            entity: 0,
          },
          found: {
            cmp: {},
            tag: {}
          }
        },

        work: {}
      }

      const defsrc = loadFile(defpath, 'def', fs, log)

      const def = await parse(opts.kind as string, defsrc, {
        file: defpath,
        // GraphQL-only inputs; ignored by the OpenAPI parser.
        graphql: {
          endpoint: opts.endpoint,
          title: model.name,
          version: (model.main as any)?.[KIT]?.info?.version,
        },
      } as any)
      const defkeys = Object.keys(def)

      log.info({
        point: 'root-keys',
        defpath,
        note: defkeys.join(', ')
      })

      // Only write the full JSON debug file when debug mode is enabled,
      // as JSON.stringify + sync write is expensive for large specs.
      if (opts.debug) {
        const fullsrc = JSON.stringify(def, null, 2)
        fs.writeFileSync(defpath + '.full.json', fullsrc)
      }

      ctx.def = def

      steps.push('parse')

      // Step: guide (derive).
      if (!ctrl.step.guide) {
        return { ok: false, steps, start, end: Date.now(), ctrl, ctx }
      }

      const guideModel = await buildGuide(ctx)
      if (null == guideModel) {
        throw new Error('Unable to build guide.')
      }

      ctx.guide = guideModel.guide

      steps.push('guide')



      // Step: transformers (transform spec and guide into core structures).
      // Early stops return the model built so far: `ctrl.step.generate = false`
      // is the documented way to build the model in memory without writing
      // files (see AGENTS.md), which requires `apimodel` in the result. The Go
      // port already returns it from every early return.
      if (!ctrl.step.transformers) {
        return {
          ok: true, steps, start, end: Date.now(), ctrl,
          guide: ctx.guide, apimodel: ctx.apimodel, ctx
        }
      }

      await topTransform(ctx)
      await entityTransform(ctx)
      await operationTransform(ctx)
      // Must precede args and field: both branch on point.kind === 'graphql'
      // and read the graphql block this stamps on.
      await graphqlTransform(ctx)
      await argsTransform(ctx)
      await selectTransform(ctx)
      await fieldTransform(ctx)
      await flowTransform(ctx)
      await flowstepTransform(ctx)
      await cleanTransform(ctx)

      steps.push('transformers')

      // Step: builders (build generated sub models).
      if (!ctrl.step.builders) {
        return {
          ok: true, steps, start, end: Date.now(), ctrl,
          guide: ctx.guide, apimodel: ctx.apimodel, ctx
        }
      }

      const builders = [
        await makeEntityBuilder(ctx),

        // TODO: move to sdkgen
        await makeFlowBuilder(ctx),
      ]

      steps.push('builders')


      // Step: generate (generate model files).
      if (!ctrl.step.generate) {
        return {
          ok: true, steps, start, end: Date.now(), ctrl,
          guide: ctx.guide, apimodel: ctx.apimodel, ctx
        }
      }

      const jostraca = Jostraca({
        now: spec.now,
        fs: () => fs,
        log,
      })

      const jmodel = {}

      const root = () => Project({ folder: '.' }, async () => {
        for (let builder of builders) {
          builder()
        }
      })

      jres = await jostraca.generate({
        // folder: Path.dirname(opts.folder as string),
        folder: opts.folder,
        model: jmodel,
        // Overwrite the generated model source (.aontu) rather than 3-way merge:
        // merging against a drifting .jostraca base silently keeps stale files
        // and can inject <<<<<<< conflict markers. Generated output is
        // model-derived and never hand-edited. See sdkgen
        // docs/explanation/regeneration-overwrite.md.
        existing: { txt: { write: true, merge: false } }
      }, root)

      const dlogs = dlog.log()
      if (0 < dlogs.length) {
        for (let dlogentry of dlogs) {
          log.debug({ point: 'generate-debug', dlogentry, note: String(dlogentry) })
        }
      }

      steps.push('generate')

      // Garbage-collect entity model files no longer derived from the def.
      // The builders only ever WRITE: a spec change that removes or renames a
      // derived entity used to leave the old <name>.aontu behind forever.
      // Runs after generate so the current set is on disk; guarded so only
      // apidef-generated files under this build's outprefix are touched.
      try {
        const kitEntity = (ctx.apimodel?.main as any)?.[KIT]?.entity || {}
        gcEntityFiles(fs, log, opts.folder as string, opts.outprefix,
          Object.keys(kitEntity))
      }
      catch (err: any) {
        log.warn({ point: 'entity-gc-failed', err, note: String(err?.message) })
      }

      const hasWarnings = 0 < warn.history.length
      const endnote =
        hasWarnings ? `PARTIAL BUILD! There were ${warn.history.length} warnings (see above).` :
          'success'
      log[hasWarnings ? 'warn' : 'info']({ point: 'generate-end', note: endnote, break: true })

      if (hasWarnings) {
        writeFileSyncWarn(warn, fs, './apidef-warnings.txt',
          warn.history.map(n => formatJSONIC(n)).join('\n\n'))
      }

      // apidef writes model source files (entity, flow, guide aontu files) into
      // .sdk/model/. Downstream actions (sdkgen, etc.) read those via
      // sdk.aontu @-includes, so voxgig-model has to re-resolve the model
      // before the post-step actions run. Signal reload whenever jostraca
      // wrote or merged any files; if nothing changed on disk,
      // voxgig-model's resolveModel cache short-circuits the re-read.
      const jfiles = jres?.files
      const reload = !!jfiles && (
        (jfiles.written?.length ?? 0) > 0 ||
        (jfiles.merged?.length ?? 0) > 0
      )

      return {
        ok: true,
        reload,
        err: null,
        start,
        end: Date.now(),
        steps,
        ctrl,
        guide: ctx.guide,
        apimodel: ctx.apimodel,
        ctx,
        jres,
      }
    }
    catch (err: any) {
      const endnote = '!! BUILD FAILED !! ' + err.message
      log.error({ point: 'generate-end', err, note: endnote, break: true })

      warn.history.push({
        point: warn.point,
        when: Date.now(),
        err,
        note: endnote
      })

      writeFileSyncWarn(warn, fs, './apidef-warnings.txt',
        warn.history.map(n => formatJSONIC(n)).join('\n\n'))

      return {
        ok: false,
        err,
        start,
        end: Date.now(),
        steps,
        ctrl,
        guide: ctx?.guide,
        apimodel: ctx?.apimodel,
        ctx,
        jres,
      }
    }
    finally {
      // Drop the per-model overrides so a subsequent generate() in
      // the same process starts with a clean utility module.
      clearCustomPlurals()
    }
  }

  return {
    generate,
  }
}


// Sniff the input format from the definition file name. GraphQL schemas use
// .graphql/.graphqls/.gql; an explicit `kind` option always wins. Introspection
// JSON must be named .graphql.json (or declared) since a bare .json could be
// either format.
function resolveKind(def?: string): DefKind {
  const name = String(def ?? '').toLowerCase()
  if (name.endsWith('.graphql') || name.endsWith('.graphqls') ||
    name.endsWith('.gql') || name.endsWith('.graphql.json')) {
    return 'GraphQL'
  }
  return 'OpenAPI'
}


ApiDef.makeBuild = async function(opts: ApiDefOptions) {
  let apidef: any = undefined

  const kind = opts.kind || resolveKind(opts.def)

  const config = {
    def: opts.def || 'no-def',
    kind: 'GraphQL' === kind ? 'graphql' : 'openapi3',
    meta: opts.meta || {},
  }

  const build = async function(model: any, build: any, _ctx: any) {

    if (null == apidef) {
      apidef = ApiDef({
        def: opts.def,
        fs: opts.fs,
        debug: opts.debug,
        folder: opts.folder,
        meta: opts.meta,
        outprefix: opts.outprefix,
        strategy: opts.strategy,
        kind,
        endpoint: opts.endpoint,
        auth: opts.auth,
        pino: build.log,
        why: opts.why,
      })
    }

    const ctrl = build.spec.buildargs?.apidef?.ctrl || {}

    return await apidef.generate({ model, build, config, ctrl })
  }

  build.step = 'pre'

  return build
}






export type {
  ApiDefOptions,
}

export type {
  PathDef,
  MethodDef,
  ServerDef,
  ServerVariableDef,
  ParameterDef,
  SchemaDef,
} from './def'

export type {
  CmpDesc,
  BasicMethodDesc,
  MethodDesc,
  MethodEntityDesc,
  EntityDesc,
  EntityPathDesc,
  PathDesc,
  OpDesc,
} from './desc'

export type {
  OpName,
  ArgKind,
  NamesCluster,
  ModelEntityRelations,
  ModelOpMap,
  ModelFieldOp,
  ModelField,
  ModelArg,
  ModelPoint,
  ModelOp,
  ModelEntity,
  Model,
  ModelEntityFlow,
  ModelEntityFlowStep,
  ModelEntityFlowStepInput,
  ModelEntityFlowStepValidator,
  ModelEntityFlowStepSpec,
} from './model'


export {
  KIT,
  ApiDef,
  gcEntityFiles,
  parse,
  formatJSONIC,
  depluralize,
  sanitizeSlug,
  slugToPascalCase,
  getModelPath,
  nom,
  VALID_CANON,
  CANON_ONE,
}
